idaes/core/util/scaling.py

```diff
@@ -49,12 +50,36 @@
 from pyomo.core import expr as EXPR
 from pyomo.common.numeric_types import native_types
 from pyomo.core.base.units_container import _PyomoUnit
+from pyomo.common.fileutils import find_library
 
 import idaes.logger as idaeslog
 
 _log = idaeslog.getLogger(__name__)
 
 
+# There appears to be a bug in the ASL which causes terminal failures
+# if you try to create multiple ASL structs with different external
+# functions in the same process. This causes pytest to crash during testing.
+# To avoid this, register all known external functions before we call
+# PyNumero.
+ext_funcs = ["cubic_roots", "general_helmholtz_external", "functions"]
+library_set = set()
+libraries = []
+
+for f in ext_funcs:
+    library = find_library(f)
+    if library not in library_set:
+        library_set.add(library)
+        libraries.append(library)
+
+if "AMPLFUNC" in os.environ:
+    env_str = "\n".join([os.environ["AMPLFUNC"], *libraries])
+else:
+    env_str = "\n".join(libraries)
+
+os.environ["AMPLFUNC"] = env_str
+
```

Comment:
Why is this in the scaling module? I could see two different approaches for implementing this workaround:
1. We don't attempt to fix this for users (until we get a bug report about it), and just add the workaround to a testing module somewhere.
2. We attempt to fix this for users.

This seems to be going with option 2. However, I would argue that, if we're trying to give users access to the workaround, we should put this in a high-level module or `__init__.py`. I say this because users could hit this issue in a wide variety of situations in which `scaling` is not imported. If we just want to narrowly fix the issue with the tests, I would put this in the test module that is triggering the bug (I believe `test_model_diagnostics`).

Reply:
The function that is being used for the Jacobian is here, hence why I moved it from the diagnostics. I think this is the only place we use PyNumero for now as well, so it would cover many user cases. The reason I did not put it in the top level `__init__.py`  is just because I am hesitant to have put code there (mostly learning from @lbianchi-lbl's warnings).

Reply:
@Robbybp I agree with your point that this code being run inconsistently depending on whether `idaes.core.util.scaling` is imported should be avoided as it could lead to very hard to debug behaviors.

As a clarification, my take is that nonstandard code being run as import-time side effects should be avoided if possible, for the reason above and others (including there not being an easy way to "opt out" or "undo" the changes if needed). However, if the choice is between "import-time side effects that are run always" vs "import-time side effects that only run sometimes", the first is IMO preferable. Additionally, since there's already a lot happening in `idaes/__init__.py`/`idaes/config.py`, I can see the argument for keeping all this type of code in one(-ish) place.

In summary, I agree with you in that `idaes/__init__.py` (or any other module that's unconditionally/consistently imported) would be the most reasonable place to put this code.